common/formats/src/format_diagnostic.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Write;
+
+use common_datavalues::DataSchemaRef;
+use common_datavalues::DataType;
+use common_datavalues::TypeDeserializerImpl;
+use common_exception::Result;
+use common_io::prelude::BufferReadExt;
+use common_io::prelude::FormatSettings;
+use common_io::prelude::MemoryReader;
+use common_io::prelude::NestedCheckpointReader;
+
+pub fn verbose_string(buf: &[u8], out: &mut String) {
+    out.push('"');
+
+    if buf.is_empty() {
+        out.push_str("<EMPTY>");
+    }
+
+    for c in buf.iter() {
+        match *c {
+            b'\0' => {
+                out.push_str("<ASCII NUL>");
+            }
+            b'\n' => {
+                out.push_str("<LINE FEED>");
+            }
+            b'\r' => out.push_str("<CARRIAGE RETURN>"),
+            b'\t' => {
+                out.push_str("<TAB>");
+            }
+            b'\\' => {
+                out.push_str("<BACKSLASH>");
+            }
+            b'"' => {
+                out.push_str("<DOUBLE QUOTE>");
+            }
+            b'\'' => {
+                out.push_str("<SINGLE QUOTE>");
+            }
+            _ => {
+                out.push(*c as char);
+            }
+        }
+    }
+    out.push('"');
+}
+
+pub trait FormatDiagnostic {
+    fn get_diagnostic_info(
+        &self,
+        buf: &[u8],
+        file_name: &Option<String>,
+        row_index: usize,
+        schema: DataSchemaRef,
+        min_accepted_rows: usize,
+        settings: FormatSettings,
+    ) -> Result<String> {
+        let mut out = String::new();
+        match file_name {
```

Comment:
maybe format!() is better(without unwrap)?

Reply:
make sense

Reply:
`format!` has extra allocation.